File: src/foam/core/config/GlobalConfigs.java
/**
 * @license
 * Copyright 2026 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package foam.core.config;

import foam.dao.DAO;
import foam.lang.X;
import java.util.Date;

/**
 * Lookup helpers for GlobalConfig. Each typed getter takes an X context,
 * an entry name, and a default value returned when the entry is missing,
 * the DAO is not registered, or the entry's type does not match.
 */
public final class GlobalConfigs {
  private GlobalConfigs() {}

  public static Object get(X x, String name) {
    GlobalConfig cfg = find_(x, name);
    return cfg == null ? null : cfg.getValue();
  }

  public static String getString(X x, String name, String defaultValue) {
    GlobalConfig cfg = find_(x, name);
    return matches_(cfg, GlobalConfigType.STRING) ? cfg.getStringValue() : defaultValue;
  }

  public static boolean getBoolean(X x, String name, boolean defaultValue) {
    GlobalConfig cfg = find_(x, name);
    return matches_(cfg, GlobalConfigType.BOOLEAN) ? cfg.getBooleanValue() : defaultValue;
  }

  public static int getInt(X x, String name, int defaultValue) {
    GlobalConfig cfg = find_(x, name);
    return matches_(cfg, GlobalConfigType.INT) ? cfg.getIntValue() : defaultValue;
  }

  public static long getLong(X x, String name, long defaultValue) {
    GlobalConfig cfg = find_(x, name);
    return matches_(cfg, GlobalConfigType.LONG) ? cfg.getLongValue() : defaultValue;
  }

  public static float getFloat(X x, String name, float defaultValue) {
    GlobalConfig cfg = find_(x, name);
    return matches_(cfg, GlobalConfigType.FLOAT) ? cfg.getFloatValue() : defaultValue;
  }

  public static double getDouble(X x, String name, double defaultValue) {
    GlobalConfig cfg = find_(x, name);
    return matches_(cfg, GlobalConfigType.DOUBLE) ? cfg.getDoubleValue() : defaultValue;
  }

  public static Date getDate(X x, String name, Date defaultValue) {
    GlobalConfig cfg = find_(x, name);
    return matches_(cfg, GlobalConfigType.DATE) ? cfg.getDateValue() : defaultValue;
  }

  public static Date getDateTime(X x, String name, Date defaultValue) {
    GlobalConfig cfg = find_(x, name);
    return matches_(cfg, GlobalConfigType.DATE_TIME) ? cfg.getDateTimeValue() : defaultValue;
  }

  private static GlobalConfig find_(X x, String name) {
    if ( x == null || name == null ) return null;
    DAO dao = (DAO) x.get("globalConfigDAO");
    if ( dao == null ) return null;
    return (GlobalConfig) dao.find(name);
  }

  private static boolean matches_(GlobalConfig cfg, GlobalConfigType type) {
    return cfg != null && cfg.getType() == type;
  }
}

File: src/foam/core/config/GlobalConfigs.js
/**
 * @license
 * Copyright 2026 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.LIB({
  name: 'foam.core.config.GlobalConfigs',

  documentation: `Lookup helpers for GlobalConfig. Each typed getter takes an
    X context, an entry name, and a default value returned when the entry is
    missing, the DAO is not registered, or the entry's type does not match.`,

  methods: [
    async function get(x, name) {
      var dao = x && x.globalConfigDAO;
      if ( ! dao || ! name ) return undefined;
      var cfg = await dao.find(name);
      return cfg ? cfg.getValue() : undefined;
    },

    async function getTyped_(x, name, type, defaultValue) {
      var dao = x && x.globalConfigDAO;
      if ( ! dao || ! name ) return defaultValue;
      var cfg = await dao.find(name);
      if ( ! cfg || cfg.type !== type ) return defaultValue;
      return cfg.getValue();
    },

    async function getString(x, name, defaultValue) {
      return this.getTyped_(x, name, foam.core.config.GlobalConfigType.STRING, defaultValue);
    },
    async function getBoolean(x, name, defaultValue) {
      return this.getTyped_(x, name, foam.core.config.GlobalConfigType.BOOLEAN, defaultValue);
    },
    async function getInt(x, name, defaultValue) {
      return this.getTyped_(x, name, foam.core.config.GlobalConfigType.INT, defaultValue);
    },
    async function getLong(x, name, defaultValue) {
      return this.getTyped_(x, name, foam.core.config.GlobalConfigType.LONG, defaultValue);
    },
    async function getFloat(x, name, defaultValue) {
      return this.getTyped_(x, name, foam.core.config.GlobalConfigType.FLOAT, defaultValue);
    },
    async function getDouble(x, name, defaultValue) {
      return this.getTyped_(x, name, foam.core.config.GlobalConfigType.DOUBLE, defaultValue);
    },
    async function getDate(x, name, defaultValue) {
      return this.getTyped_(x, name, foam.core.config.GlobalConfigType.DATE, defaultValue);
    },
    async function getDateTime(x, name, defaultValue) {
      return this.getTyped_(x, name, foam.core.config.GlobalConfigType.DATE_TIME, defaultValue);
    }
  ]
});
